BUG: np.squeeze() now respects older API axis expectation

tylerjereddy opened the pull request:
I tried to patch this issue; the descriptions of the fix make it sound a little obscure, but to be fair it is a little bit tricky to think about since it seems to be about respecting an old API and preventing a silent success / enforcement of new API.

* Fixes Issue #10779 by removing the interception of
an otherwise normal Exception when an object implemented
with the expectation that squeeze() does not accept an
axis argument receives an axis argument

* Added unit tests that enforce respect for the old API
expectation in objects, and ensure that silent success
(or forced usage of the new API on objects) is no longer
the case

* Updated compatibility notes to explain this change

eric-wieser commented on numpy/core/tests/test_regression.py:
```diff
@@ -1675,6 +1675,66 @@ def test_squeeze_contiguous(self):
+
+        # the new API knows how to handle the axis
+        # argument and will return a ValueError if
+        # attempting to squeeze a non-empty axis
```
`non-empty` is not correct, since that implies not 0. What you mean is not 1.

tylerjereddy committed:
BUG: np.squeeze() now respects older API axis expectation

* Fixes Issue #10779 by removing the interception of
an otherwise normal Exception when an object implemented
with the expectation that squeeze() does not accept an
axis argument receives an axis argument

* Added unit tests that enforce respect for the old API
expectation in objects, and ensure that silent success
(or forced usage of the new API on objects) is no longer
the case

* Updated compatibility notes to explain this change

charris commented:
Thanks @tylerjereddy .

cdeil commented:
This change broke our code:
https://travis-ci.org/gammapy/gammapy/jobs/363544680#L3839
We pass a list of int for axis and now get:
```
 TypeError: 'list' object cannot be interpreted as an integer
```
Should list of int be allowed here?
If no, I think we can just change to tuple of int and it should work again?
I'll try that now.

eric-wieser commented:
Your code was already broken - the list you were passing was being completely ignored. To keep the behavior you had before, you can call it with no arguments as `squeeze()`

cdeil commented:
@eric-wieser - In that case: thanks for catching the issue for us!
I have to familiarise myself with that part of code first, I'll have a look tomorrow.
